crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -2032,11 +2286,184 @@ mod test_shielded_wallet {
+                        token: native_token.clone(),
+                        denom: native_token_denom,
+                        position: MaspDigitPos::Zero,
+                        epoch: Some(MaspEpoch::new(2)),
+                    }.encode().unwrap(),
+                    1 + reward_rate,
+                );
+                context.add_conversions(
+                    AssetData {
+                        token: native_token.clone(),
+                        denom: native_token_denom,
+                        position: MaspDigitPos::Zero,
+                        epoch: Some(MaspEpoch::new(1)),
+                    },
+                    (
+                        native_token.clone(),
+                        native_token_denom,
+                        MaspDigitPos::Zero,
+                        MaspEpoch::new(1),
+                        conv,
+                        MerklePath::from_path(vec![], 0),
+                    )
+                );
+                let vk = arbitrary_vk();
+                let pa = arbitrary_pa();
+                for epoch in [0, 2] {
+                    let asset_data = AssetData {
+                        token: native_token.clone(),
+                        denom: native_token_denom,
+                        position: MaspDigitPos::Zero,
+                        epoch: Some(MaspEpoch::new(epoch)),
+                    };
+
+                    wallet.add_note(
+                        create_note(asset_data, principal, pa),
+                        vk,
+                    );
+                }
+
+                // If this flag is clear we test that the estimation logic can
+                // hanlde rewards for the native asset coming only from the current epoch,
+                // i.e. for which there are no conversions yet.
+                if shield_at_previous_epoch {
+                    let asset_data = AssetData {
+                        token: native_token.clone(),
+                        denom: native_token_denom,
+                        position: MaspDigitPos::Zero,
+                        epoch: Some(MaspEpoch::new(1)),
+                    };
+
+                    wallet.add_note(
+                        create_note(asset_data, principal, pa),
+                        vk,
+                    );
+                }
+
+                let balance = wallet
+                    .compute_shielded_balance(&vk)
+                    .await
+                    .unwrap()
+                    .unwrap_or_else(ValueSum::zero);
+                let rewards_est = wallet.estimate_next_epoch_rewards(&context, &balance).await.expect("Test failed");
+                if shield_at_previous_epoch {
+                    assert_eq!(
+                        rewards_est,
+                        DenominatedAmount::native(
+                            Amount::from_masp_denominated_i128(
+                                (2 + reward_rate) * reward_rate * i128::from(principal),
```

Comment:
Another mysterious `2+`. Maybe write out the formula's meaning in a comment.

Reply:
Same answer as before, only difference is that rewards for the native token compound, so we need to add 2 instead of 1 to the multiplier this time

Reply:
Same as before, but in this case we should also account for the compounding of rewards when the native token itself accrues rewards, I've added a comment to the formula